src/components/Launcher.jsx

```diff
@@ -0,0 +1,188 @@
+import React, { Component } from 'react'
+import { translate } from '../lib/I18n'
+import Autosuggest from 'react-autosuggest';
+import { rawIntent } from '../lib/intents'
+
+const INTENT_VERB = 'OPEN'
+const INTENT_DOCTYPE = 'io.cozy.suggestions'
+
+class Launcher extends Component {
+  constructor (props) {
+    super(props)
+    
+    this.state = {
+      query: '',
+      suggestions: []
+    }
+    
+    this.sourcesNode = null
+    this.sources = []
+  }
+  
+  componentDidMount () {
+    // The launcher has one or more sources that provide suggestions. These sources are iframes into other apps, provied by thee intent system.
+    // Since we need to call the sources whenever the query changes, we are taking manual control over the intent process.
+    window.addEventListener('message', this.onMessageFromSource)
+    
+    rawIntent(INTENT_VERB, INTENT_DOCTYPE)
+    .then(intent => {
+      const { services } = intent.attributes
+      if (!services) return
+      
+      services.forEach(service => {
+        // we're injecting the iframes, but they will stay invisible
+        const iframe = document.createElement('iframe')
+        const url = service.href
+        iframe.setAttribute('src', url)
+        iframe.style.display = 'none'
+        this.sourcesNode.appendChild(iframe)
+        const serviceOrigin = url.split('/', 3).join('/')
+        
+        this.sources.push({
+          slug: service.slug,     // can be used to show where a suggestion comes from
+          origin: serviceOrigin,
+          id: intent._id,
+          ready: false,
+          window: null,           // will hold a reference to the window we're sending messages to
+          resolve: null           // a reference to a function to call when the source sends suggestions
+        })
+      })
+    })
+  }
+  
+  onMessageFromSource = (event) => {
```

Comment:
It does multiple things, depending on the type of message, so it's not really possible to express the code's intention in the function name. At best it would be `handleIntentMessage` or something equally not-very-informative.
In this case the `onSomeEvent` form is also in line with the rest of the component (`onSuggestionsFetchRequested` and all from React AutoSuggest).